Refs #31617 -- Added an id for helptext in admin forms.

smithdc1 opened the pull request:
Refs Admin Accessibility Ticket https://code.djangoproject.com/ticket/31617

This adds an `id` to the helptext in the admin and should be merged before https://github.com/django/django/pull/15246

If I stack this pr and 15246 then we see the helptext will be associated with the input. 

![image](https://user-images.githubusercontent.com/39445562/147889992-59e68a9b-9275-4e9d-aa90-10d4143344a3.png)


Line 5216 in `tests/admin_views/tests.py` doesn't change because dates are `multiwidgets` and therefore they don't get an `id`. Likely that needs some more thought, but I think that can be addressed separately. Split Date/time widgets need work more generally, likely need a fieldset and a legend etc.
https://github.com/django/django/blob/4400d8568ad5695c46e8de45635a82a27a26b871/django/forms/widgets.py#L852

felixxm commented:
@smithdc1 Thanks :+1: I found a few more places:

- `django/contrib/admin/templates/registration/password_change_form.html`: `<div class="help">{{ form.new_password1.help_text|safe }}</div>`,
- `django/contrib/admin/templates/registration/password_change_form.html`: `<div class="help">{{ form.new_password2.help_text|safe }}</div>`,
- `django/contrib/admin/templates/admin/auth/user/change_password.html`: `<div class="help">{{ form.password1.help_text|safe }}</div>`,
- `django/contrib/admin/templates/admin/auth/user/change_password.html`: `<div class="help">{{ form.password2.help_text|safe }}</div>`,
- `django/contrib/admin/templates/admin/search_form.html`: `<div class="help">{{ cl.search_help_text }}</div>`

smithdc1 commented:
>  found a few more places:

Thanks :-)

The first 4 are fairly straight forward, and I've assumed that folk have `id`s for their password inputs. 

The last one is a little different as the input is in the template so I've added that `aria-describedby` (not labelled) here. I've added as a single commit (to be squashed) but maybe the change to the input should be a different commit or part of #15246?

smithdc1 commented:
>  found a few more places:

Thanks :-)

The first 4 are fairly straight forward, and I've assumed that folk have `id`s for their password inputs. 

The last one is a little different as the input is in the template so I've added that `aria-describedby` (not labelled) here. I've added as a single commit (to be squashed) but maybe the change to the input should be a different commit or part of #15246?

smithdc1 commented:
> Line 5216 in tests/admin_views/tests.py doesn't change because dates are multiwidgets and therefore they don't get an id. 

hmm, should be using `auto_id` instead :thinking:
https://github.com/django/django/blob/205f67cd5bea242fac6b0380fa26570505d07dcb/django/forms/boundfield.py#L219

I need to think more about this.

smithdc1 commented:
Hi @felixxm

Can I get your thoughts please? 

Currently I've been using `id_for_label` and this gives us `aria-describedby` on the correct widgets. That is everything except where there's multiple inputs involved, e.g. MultiWidget as mentioned above but also checkbox and radios won't get a `aria-describedby` as my patches currently stand. Instead for these it seems we should be adding the `aria-describedby` to the `fieldset` See [gds example](https://design-system.service.gov.uk/components/radios/hint/index.html). Specifics aside, my key point here is that it's wrong to add an `aria-describedby` from every radio (say) to the field's `help_text`. Further, for these widgets we'd need to try and use `auto_id` as their labels don't have an `id`. 

I'm wondering instead if widgets should learn a `use_fieldset` attribute (name can be improved). This will be `False` by default and `True` for radio/multi-checkboxes etc. We can then use `auto_id` instead and this new attribute. I think this approach has the following benefits:

- We can simplify the fieldset logic in the proposed `<div>` template to use this new attribute, rather than listing out widget names.
- Improved consistency through use of `auto_id` everywhere we're going to associate `help_text` with an input. (What's the `label` got to do with `help_text` anyway?). We'll guard against applying the `aria-describedby` to the wrong widgets using the new attribute.
- Folk with custom widgets will be able to get further by using the attribute on their widget. 
- Associating the `help_text` with the input will not be impacted by folk customising `id_for_label`. 

I guess my main concern about this approach is we're adding _another_ attribute (setting?) to widgets.

felixxm commented:
> The first 4 are fairly straight forward, and I've assumed that folk have `id`s for their password inputs.

I don't think we can safely assume this :think I would add `{% if form.password1.id_for_label %} ...` etc.

felixxm commented:
> [ticket-31617](https://code.djangoproject.com/ticket/31617) is already wide and it's getting harder and harder to follow all threads, opinions, and decisions :|

OK, we have ticket-32819 :facepalm:

smithdc1 commented:
> getting harder and harder to follow all threads, opinions, and decisions :|

Completely agree, especially when I'm trying to push on with related tickets at the same time. 

I'll think today about how to structure the order of these changes so they are easier to follow.

carltongibson commented:
Hey @smithdc1 — so a quick look at this: it looks good. 

> ...harder to follow... 

What with here, https://github.com/django/django/pull/15246, and https://github.com/django/django/pull/15241 — it's like _Argh!_ 😃 

Is this one ready to go. Getting this in lets us look again at #15246 I think. 

🏅

smithdc1 commented:
Yes. This one should be good to go, albeit I hadn't refreshed my memory enough of this to put it in the review queue more formally. 

P.s. nearly there with my thoughts on the 'div' template. Hopefully get something pushed later with a proposal for consideration.

carltongibson reviewed (approved):
OK, LGTM. Thanks @smithdc1! Onwards... 😄
